File: app/src/main/java/com/github/goldy1992/mp3player/client/MediaBrowserAdapter.java
package com.github.goldy1992.mp3player.client;

import android.os.Bundle;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.session.MediaSessionCompat;

import com.github.goldy1992.mp3player.client.callbacks.connection.MyConnectionCallback;
import com.github.goldy1992.mp3player.client.callbacks.search.MySearchCallback;
import com.github.goldy1992.mp3player.client.callbacks.search.SearchResultListener;
import com.github.goldy1992.mp3player.client.callbacks.subscription.MediaIdSubscriptionCallback;
import com.github.goldy1992.mp3player.dagger.scopes.ComponentScope;

import javax.inject.Inject;

@ComponentScope
public class MediaBrowserAdapter {

    private static final String LOG_TAG = "MDIA_BRWSR_ADPTR";
    private MediaBrowserCompat mediaBrowser;
    private MyConnectionCallback connectionCallback;
    private MediaIdSubscriptionCallback mySubscriptionCallback;
    private MySearchCallback mySearchCallback;

    @Inject
    public MediaBrowserAdapter(MediaBrowserCompat mediaBrowser,
                               MyConnectionCallback myConnectionCallback,
                               MediaIdSubscriptionCallback mySubscriptionCallback,
                               MySearchCallback mySearchCallback) {
        this.mediaBrowser = mediaBrowser;
        this.connectionCallback = myConnectionCallback;
        this.mySubscriptionCallback = mySubscriptionCallback;
        this.mySearchCallback = mySearchCallback;
    }

    public void init() {
        // Create MediaBrowserServiceCompat
        mediaBrowser.connect();
        //Log.i(LOG_TAG, "calling connect");
    }

    /**
     * Disconnects from the media browser service
     */
    public void disconnect() {
        mediaBrowser.disconnect();
    }

    public void search(String query, Bundle extras) {
        mediaBrowser.search(query, extras, mySearchCallback);
    }
    /**
     * Connects to the media browser service
     */
    public void connect() {
        mediaBrowser.connect();
    }

    /**
     * subscribes to a MediaItem via a libraryRequest. The id of the libraryRequest will be used for the parent
     * ID when communicating with the MediaPlaybackService.
     * @param id the id of the media item to be subscribed to
     */
    public void subscribe(String id) {
        mediaBrowser.subscribe(id, mySubscriptionCallback);
    }

    public void subscribeToRoot() {
        mediaBrowser.subscribe(getRootId(), mySubscriptionCallback);
    }

    public MediaSessionCompat.Token getMediaSessionToken() {
        return mediaBrowser.getSessionToken();
    }

    public String getRootId() {
        return mediaBrowser.getRoot();
    }

    public boolean isConnected() {
        return mediaBrowser != null && mediaBrowser.isConnected();
    }

    public void registerRootListener(MediaBrowserResponseListener mediaBrowserResponseListener) {
        mySubscriptionCallback.registerMediaBrowserResponseListener(getRootId(), mediaBrowserResponseListener);
    }

    public void registerListener(String parentId, MediaBrowserResponseListener mediaBrowserResponseListener) {
        mySubscriptionCallback.registerMediaBrowserResponseListener(parentId, mediaBrowserResponseListener);
    }

    public void registerSearchResultListener(SearchResultListener searchResultListener) {
        this.mySearchCallback.registerSearchResultListener(searchResultListener);
    }

    public boolean unregisterSearchResultListener(SearchResultListener searchResultListener) {
        return this.mySearchCallback.unregisterSearchResultListener(searchResultListener);
    }

    public MediaIdSubscriptionCallback getMySubscriptionCallback() {
        return mySubscriptionCallback;
    }

    public MyConnectionCallback getConnectionCallback() {
        return connectionCallback;
    }
}

File: client/src/main/java/com/github/goldy1992/mp3player/client/MediaBrowserAdapter.kt
package com.github.goldy1992.mp3player.client

import android.os.Bundle
import android.util.Log
import androidx.concurrent.futures.await
import androidx.media3.common.MediaItem
import androidx.media3.session.LibraryResult
import androidx.media3.session.MediaBrowser
import androidx.media3.session.MediaLibraryService
import com.github.goldy1992.mp3player.commons.Constants.PACKAGE_NAME
import com.github.goldy1992.mp3player.commons.Constants.PACKAGE_NAME_KEY
import com.github.goldy1992.mp3player.commons.LogTagger
import com.github.goldy1992.mp3player.commons.MainDispatcher
import com.google.common.collect.ImmutableList
import com.google.common.util.concurrent.ListenableFuture
import dagger.hilt.android.scopes.ActivityRetainedScoped
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.apache.commons.lang3.StringUtils.isEmpty
import javax.inject.Inject

@ActivityRetainedScoped
open class MediaBrowserAdapter

    @Inject
    constructor(private val mediaBrowserLF : ListenableFuture<MediaBrowser>,
    private val scope: CoroutineScope,
    @MainDispatcher private val mainDispatcher: CoroutineDispatcher) : LogTagger {

    companion object {
        private fun getDefaultLibraryParams() : MediaLibraryService.LibraryParams {
            return MediaLibraryService.LibraryParams.Builder().build()
        }
    }

    open suspend fun search(query: String, extras: Bundle) {
        if (isEmpty(query)) {
            Log.w(logTag(), "Null or empty search query seen")
        }
        else {
            val params = MediaLibraryService.LibraryParams.Builder().setExtras(extras).build()
            mediaBrowserLF.await().search(query, params)
        }

    }

    open suspend fun getSearchResults(query: String, page : Int = 0, pageSize : Int = 20) : ImmutableList<MediaItem> {
        val result : LibraryResult<ImmutableList<MediaItem>> =
            mediaBrowserLF.await().getSearchResult(query, page, pageSize, getDefaultLibraryParams()).await()
        return result.value ?: ImmutableList.of()
    }

    /**
     * subscribes to a MediaItem via a libraryRequest. The id of the libraryRequest will be used for the parent
     * ID when communicating with the MediaPlaybackService.
     * @param id the id of the media item to be subscribed to
     */
    open suspend fun subscribe(id: String) {
        mediaBrowserLF.await().subscribe(id, getDefaultLibraryParams())
    }

    open suspend fun getLibraryRoot() : MediaItem {
        val args = Bundle()
        args.putString(PACKAGE_NAME_KEY, PACKAGE_NAME)
        val params = MediaLibraryService.LibraryParams.Builder().setExtras(args).build()
        val result = mediaBrowserLF.await().getLibraryRoot(params).await()
        return result.value ?: MediaItem.EMPTY
    }

    suspend fun getChildren(parentId : String,
                            @androidx.annotation.IntRange(from = 0) page : Int = 0,
                            @androidx.annotation.IntRange(from = 1) pageSize : Int = 20,
                            params : MediaLibraryService.LibraryParams = MediaLibraryService.LibraryParams.Builder().build()
    ) : List<MediaItem> {
        val children : LibraryResult<ImmutableList<MediaItem>> = mediaBrowserLF.await().getChildren(parentId, page, pageSize, params).await()
        return children.value?.toList() ?: emptyList()
    }

    override fun logTag(): String {
        return "MDIA_BRWSR_ADPTR"
    }

}